### PR TITLE
Audit revision

### PR DESCRIPTION
**Pull request by bbenligiray**
Since we decided to redeploy for v0.4, I figured it's a good idea to implement the minor fixes. You may want to go commit by commit.

### CODE REVIEW

File: packages/airnode-protocol/contracts/rrp/AirnodeRrp.sol
Comment (bbenligiray) [inline packages/airnode-protocol/contracts/rrp/AirnodeRrp.sol:86]: We agreed that this is redundant in the current state in the post-audit communications

File: packages/airnode-protocol/contracts/whitelist/WhitelistRoles.sol
Comment (bbenligiray) [inline packages/airnode-protocol/contracts/whitelist/WhitelistRoles.sol:22]: This is to address API3-08-2. I tried to come up with a different than "admin" here, but I think it's actually apt and some additional documentation is probably enough to clear away potential confusion.

File: packages/airnode-protocol/test/rrp/requesters/RrpBeaconServer.sol.js
Comment (bbenligiray) [inline packages/airnode-protocol/test/rrp/requesters/RrpBeaconServer.sol.js:68]: After implementing the `bytes(description).length > 0` check I found out that these numbers were being sent as empty

**Review by Ashar2shahid — COMMENTED** [review]
Some comments

File: packages/airnode-protocol/hardhat.config.js
Comment (aquarat) [reply packages/airnode-protocol/hardhat.config.js:52]: It configures the number of runs the solidity compiler's optimiser does, you can read about it here: [Solidity Optimizer](https://docs.soliditylang.org/en/v0.8.10/internals/optimizer.html). It determines a trade-off between deployed code size and execution complexity.

File: packages/airnode-protocol/contracts/rrp/TemplateUtils.sol
Comment (aquarat) [reply packages/airnode-protocol/contracts/rrp/TemplateUtils.sol:65]: It's surprising to me that the solidity compiler wouldn't automatically optimise out something like this - crazy.

File: packages/airnode-protocol/contracts/rrp/WithdrawalUtils.sol
Comment (aquarat) [reply packages/airnode-protocol/contracts/rrp/WithdrawalUtils.sol:31]: If a withdrawal succeeds, the source wallet will be close to empty (EIP-1559) or completely empty (legacy gas pricing), in which case future attempts to withdraw will fail as Airnode checks the balance before attempting to process a withdrawal.

File: packages/airnode-protocol/contracts/rrp/AirnodeRrp.sol
Comment (aquarat) [reply packages/airnode-protocol/contracts/rrp/AirnodeRrp.sol:63]: I suspect it isn't possible to do it on-chain with the current `calldata` as it lacks the Airnode's xpub, but besides that it would also add additional computational complexity on-chain without any benefit (if the request is invalid, the node won't respond and the request will drift out of the logs retrieval window).
I personally don't think it's necessary to detail this in the comment/inline docs as if you're reading at this level you should be able to figure it out (if it was a question in the first place).

File: packages/airnode-protocol/contracts/access-control-registry/AccessControlRegistry.sol
Comment (aquarat) [reply packages/airnode-protocol/contracts/access-control-registry/AccessControlRegistry.sol:75]: The `description` is used in `_deriveRole` and is keccak-hashed, so having an empty description would increase the risk of a hash collision/the description acts as a "[salt](https://en.wikipedia.org/wiki/Salt_(cryptography))".

File: packages/airnode-protocol/contracts/rrp/TemplateUtils.sol
Comment (aquarat) [reply packages/airnode-protocol/contracts/rrp/TemplateUtils.sol:65]: I reproduced this locally and amazingly caching a bytes32 value from a mapping saved around 3407 gas units for a 0 optimisation runs value and 1743 gas units for a 200 optimisation runs value. For a 2000 optimisation runs value the cost delta increased to 1788 - so excessive optimiser runs isn't just diminishing returns, it can in fact become more expensive.

**Comment by aquarat** [comment]
LGTM :)

File: packages/airnode-protocol/contracts/rrp/WithdrawalUtils.sol
Comment (bbenligiray) [reply packages/airnode-protocol/contracts/rrp/WithdrawalUtils.sol:31]: `require` gas cost is typically not a consideration. The main reason I'm not doing the same check in `fulfillWithdrawal()` is that https://github.com/api3dao/airnode/blob/322e91b620da13fcd02b2667875c333a3c8e0855/packages/airnode-protocol/contracts/rrp/WithdrawalUtils.sol#L65-L69 already does it implicitly, i.e., it is guaranteed that this will revert if `airnode` is `address(0)`

File: packages/airnode-protocol/contracts/rrp/AirnodeRrp.sol
Comment (bbenligiray) [reply packages/airnode-protocol/contracts/rrp/AirnodeRrp.sol]: I extended the docs (https://github.com/api3dao/airnode/pull/748/commits/085c68aa0f3627bb1d453ef1334646734632dc22) but the exact why is a subjective choice (as in you may have as easily require `fulfillAddress` to belong to a contract at request or fulfillment time)

File: packages/airnode-protocol/contracts/rrp/WithdrawalUtils.sol
Comment (bbenligiray) [reply packages/airnode-protocol/contracts/rrp/WithdrawalUtils.sol:31]: Documented here https://github.com/api3dao/airnode/pull/748/commits/39014e3be1289533d58a57ead51a41ec73d76df2

File: packages/airnode-protocol/contracts/access-control-registry/AccessControlRegistry.sol
Comment (bbenligiray) [reply packages/airnode-protocol/contracts/access-control-registry/AccessControlRegistry.sol:75]: This is a sanity check, which prove its worth already in https://github.com/api3dao/airnode/pull/748#discussion_r771626991
One thing to note is that in some cases we do want a hash collision. Say I have two authorizer contracts and I want them to share their roles, then I would deploy them with the same admin role description.

**Comment by bbenligiray** [comment]
@aquarat We already have a script for it here https://github.com/api3dao/airnode/blob/audit-revision/packages/airnode-protocol/package.json#L30 I didn't check what Yul does though so go ahead pls.

**Comment by aquarat** [comment]
> @aquarat We already have a script for it here https://github.com/api3dao/airnode/blob/audit-revision/packages/airnode-protocol/package.json#L30 I didn't check what Yul does though so go ahead pls.

Ah great, there is a saving for enabling Yul, but it is negligible (8 gas cost units against 365810 units).

File: packages/airnode-protocol/hardhat.config.js
Comment (bbenligiray) [reply packages/airnode-protocol/hardhat.config.js]: Increased to 1000 after Aaron's analysis

**Comment by bbenligiray** [comment]
Will merge after an approval :pray:

**Review by aquarat — APPROVED** [review]
LGTM :1st_place_medal:

File: packages/airnode-protocol/contracts/rrp/WithdrawalUtils.sol
Comment (Ashar2shahid) [reply packages/airnode-protocol/contracts/rrp/WithdrawalUtils.sol:31]: ah yeah, that makes sense

**Comment by Ashar2shahid** [comment]
Just went through all the comments, LGTM as well

**Pull request merged**
